[SVM GC] Add Survivor Space to SVM GC

tbzhang opened the pull request:
This patch will use the multi-survivors for SVM GC to reduce the frequency of full GC.
In the micro benchmark it could reduce the frequently of full GC(the full GC times could change from ~17/260 to ~9/250)
And in the real workload in Alibaba, it could also reduce the frequency of full GC(the full GC times could change from ~30/100 to ~1/200)

christianwimmer commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -140,4 +216,135 @@ boolean slowlyFindPointer(Pointer p) {
+        trace.string(", nofSurvivors=").unsigned(totalNofSurvivorChunks)
+                .string(", maxSurvivors=").unsigned(HeapPolicy.getMaxNofSurvivorChunks());
+        if (originalSpace.isYoungSpace()
+                && this.totalNofSurvivorChunks <= HeapPolicy.getMaxNofSurvivorChunks().rawValue()) {
```
This seems to be a quite ineffective policy to decide whether to keep an object in the young generation. When the threshold is crossed while a collection is running, suddenly all remaining objects (even the ones from eden) are ending up in the old generation.

I think we should instead collect some statistics about survivor space sizes (by age) and promotion rates, and at the beginning of a collection estimate how many survivor spaces should be used. That works because the chunked heap model does not impose strict size limits on the survivor spaces. If the estimate was too optimistic, the young generation is just a bit larger than expected after a collection, but no other harm is done. The next young generation collection will then use a smaller limit for survivor spaces, i.e., promote earlier to the old generation.

And if the promotion statistics say that the promotion rate from space `n` to `n + 1` is very high (say > 80% or so), then it is also not beneficial to use space `n + 1` anymore and instead objects should be promoted to the old generation.

All of that can and should be abstracted in a separate policy class so that it is easy to experiment with different promotion strategies.

christianwimmer commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -41,6 +43,11 @@
+    private final SurvivorSpace[] survivorFromSpaces;
+    private final SurvivorSpace[] survivorToSpaces;
+    private final GreyObjectsWalker[] survivorGreyObjectsWalkers;
+    static final int MAX_SURVIVORS = 15;
```
This can be a `HostedOption` instead of a constant in source code.

tbzhang replied:
done

christianwimmer commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapPolicy.java:
```diff
@@ -186,6 +186,31 @@ private static UnsignedWord getAllocationBeforePhysicalMemorySize() {
 
+    /* Survivor space configuration */
+
+    public static boolean useSurvivors() {
```
I do not see a reason to keep the "old" mode without survivor spaces. Do you have benchmarks or use cases where performance is better without survivor spaces?

tbzhang replied:
done, we keep the "old" mode because we want to offer a switchable option, there is no regression of the "old" mode in our test

christianwimmer commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/SurvivorSpace.java:
```diff
@@ -0,0 +1,45 @@
+    final private boolean isFrom;
+
+    /* Index of survivor space. */
+    final private int survivorIndex;
```
Isn't this usually just called `age`?

christianwimmer commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -86,18 +120,60 @@ public final Space getSpace() {
         return space;
     }
 
+    private final SurvivorSpace getSurvivorToSpaceAt(int index) {
```
I'm generally fine to not have accessor methods for private or project-private fields, to keep the code shorter. In this particular case I think the assertion is also not overly useful so it can be dropped too.

tbzhang replied:
done

christianwimmer commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -41,6 +43,11 @@
+    private final SurvivorSpace[] survivorToSpaces;
+    private final GreyObjectsWalker[] survivorGreyObjectsWalkers;
+    static final int MAX_SURVIVORS = 15;
+    private int totalNofSurvivorChunks = 0;
```
As mentioned in a longer comment elsewhere, I  don't think "counting survivor chunks" is a good policy.

tbzhang commented:
> Thanks for your contribution! Tenuring in the young generation is an important feature that was so far missing.
> 
> I added several more detailed comments, but here is a high-level summary:
> 
> * The `age` information should be stored in each `HeapChunk.Header` for fast access (and it should be consistently called `age`).
> * Objects in aligned and unaligned chunks should be handled the same way, i.e., large arrays should use survivor spaces too.
> * The policy how many survivor spaces to use should be factored out in a policy, and the default policy should be based on historic promotion rates.

Hi Christian, thank you for the suggestions!

It makes sense to me that the `age` information should be stored in `HeapChunk.Header`, will do that and update this PR soon.

I didn't see many unaligned chunks in our scenarios, that’s why unsigned chunks are not supported in the original patch. But indeed adding `age` to unaligned chunks will make this a more complete solution, will do that.

About the policy of survivors, although current implementation is a little rigid, we still see significant improvements after applying it to our workloads. How about optimizing policy in another pull-request? then I can make this patch more focused and also easier for code review.

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapImpl.java:
```diff
@@ -268,16 +271,77 @@ final Space getAllocationSpace() {
-        final Object result = oldGen.promoteObject(original);
+        Object result;
+        if (HeapPolicy.useSurvivors() && !getGCImpl().isCompleteCollection()
+                && ObjectHeaderImpl.getObjectHeaderImpl().isAlignedObject(original)) {
```
Yes, we will update this soon.

tbzhang replied:
done, treat the unaligned object in the same way

christianwimmer commented:
> About the policy of survivors, although current implementation is a little rigid, we still see significant improvements after applying it to our workloads. How about optimizing policy in another pull-request? then I can make this patch more focused and also easier for code review.

I'm fine with having a single simple policy in the initial PR, and refine later. But then the first policy should really be the simplest one: always use all tenuring spaces (with a sensible default of 2 or so) without any size checks. That can then serve as a baseline for all potentially better policies.

tbzhang commented:
> > About the policy of survivors, although current implementation is a little rigid, we still see significant improvements after applying it to our workloads. How about optimizing policy in another pull-request? then I can make this patch more focused and also easier for code review.
> 
> I'm fine with having a single simple policy in the initial PR, and refine later. But then the first policy should really be the simplest one: always use all tenuring spaces (with a sensible default of 2 or so) without any size checks. That can then serve as a baseline for all potentially better policies.

I agree with that simplest policy as the baseline, will do that then update this PR.

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/SurvivorSpace.java:
```diff
@@ -0,0 +1,45 @@
+ * The survivor space for young generation, only works with -XX:+UseSurvivors.
+ *
+ */
+public class SurvivorSpace extends Space {
```
done, move this fields to Space class

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -140,4 +216,135 @@ boolean slowlyFindPointer(Pointer p) {
+        this.totalNofSurvivorChunks += 1;
+    }
+
+    public boolean shouldPromoteToYoung(Space originalSpace) {
```
done, the age information adding to `HeapChunk.Header`

christianwimmer commented:
Thanks for the changes!  @christianhaeubl please also have a detailed look at the PR.

christianhaeubl commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/GreyToBlackObjRefVisitor.java:
```diff
@@ -140,6 +160,7 @@ public boolean visitObjectReferenceInline(final Pointer objRef, final int innerO
         } else {
             getCounters().noteUnmodifiedReference();
         }
+        HeapImpl.getHeapImpl().dirtyCardIfNecessary(holderObject, objRef, copy);
```
Please add a comment that describes why it is crucial to dirty the card even though the reference was not updated (i.e., `copy == obj`).

tbzhang replied:
done, it doesn't need dirty the card when `copy == obj`, `dirtyCardIfNecessary ` should only called when the reference was updated

christianhaeubl replied:
Are you sure about that? What if an unaligned chunk is promoted? In that case, the object won't move, only the chunk will get assigned to a different space.

tbzhang replied:
Yes, you are right, the unaligned object promotion will not update the reference, I add a comment to describe this

tbzhang commented:
> Thanks for your contribution! I added a couple of comments. Most of them are minor requests for cleanups.
> 
> However, I think that there is one larger issue remaining: the current implementation does not limit the young generation size (`HeapPolicy.getMaximumYoungGenerationSize()`) correctly. Instead, the size limit for the young generation rather acts as a size limit for eden, see `HeapPolicy.Sometimes.maybeCauseCollection()`.

Yes, the usage survivor space should be considered when calling `HeapPolicy.Sometimes.maybeCauseCollection()`.
So, I change the `maybeCauseCollection` and keep the `getMaximumYoungGenerationSize` as the original version

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/GCImpl.java:
```diff
@@ -469,6 +473,11 @@ private boolean checkIfOutOfMemory() {
         final UnsignedWord allowed = HeapPolicy.getMaximumHeapSize();
         /* Only the old generation has objects in it because the young generation is empty. */
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapImpl.java:
```diff
@@ -271,13 +271,60 @@ final Space getAllocationSpace() {
+            return;
+        }
+
+        final ObjectHeaderImpl ohi = HeapImpl.getHeapImpl().getObjectHeaderImpl();
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapImpl.java:
```diff
@@ -271,13 +271,60 @@ final Space getAllocationSpace() {
+        trace.string(", objRef: ").hex(objRef);
+        trace.string(", object: ").hex(Word.objectToUntrackedPointer(object)).newline();
+
+        if (holderObject == null || HeapImpl.getHeapImpl().getGCImpl().isCompleteCollection()) {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapImpl.java:
```diff
@@ -271,13 +271,60 @@ final Space getAllocationSpace() {
+        }
+
+        final ObjectHeaderImpl ohi = HeapImpl.getHeapImpl().getObjectHeaderImpl();
+        if (ohi.isAlignedObject(object)) {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/OldGeneration.java:
```diff
@@ -107,6 +109,10 @@ public Object promoteObject(Object original) {
     void releaseSpaces() {
         /* Release any spaces associated with this generation after a collection. */
         getFromSpace().release();
+        if (!HeapImpl.getHeapImpl().getGCImpl().isCompleteCollection()) {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/Space.java:
```diff
@@ -69,6 +69,15 @@ public Accounting getAccounting() {
     protected final String name;
 
+    /** Flag specifying if this is a survivor space. */
+    private boolean isSurvivorSpace;
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/Space.java:
```diff
@@ -136,6 +156,37 @@ final boolean isYoungSpace() {
+        return age;
+    }
+
+    void setAge(int age) {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/Space.java:
```diff
@@ -69,6 +69,15 @@ public Accounting getAccounting() {
+    private boolean isSurvivorSpace;
+
+    /** The age of this Space. */
+    private int age;
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -41,6 +43,9 @@
 
     // Final State.
     private final Space space;
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -53,6 +58,15 @@
+        this.survivorToSpaces = new Space[maxSurvivors];
+        this.survivorGreyObjectsWalkers = new GreyObjectsWalker[maxSurvivors];
+        for (int i = 0; i < maxSurvivors; i++) {
+            this.survivorFromSpaces[i] = new Space("Survivor-" + i + " From", true, i);
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/Space.java:
```diff
@@ -69,6 +69,15 @@ public Accounting getAccounting() {
+    private int age;
+
+    /** Flag specifying if this is a From space. */
+    private boolean isFrom;
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapImpl.java:
```diff
@@ -271,13 +271,60 @@ final Space getAllocationSpace() {
+        }
+
+        final UnsignedWord objectHeader = ObjectHeaderImpl.readHeaderFromObject(holderObject);
+        if (ohi.isAlignedObject(holderObject)) {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapPolicy.java:
```diff
@@ -186,6 +186,25 @@ private static UnsignedWord getAllocationBeforePhysicalMemorySize() {
+
+
+    public static int getMaxSurvivors() {
+        final int maxSurvivors = HeapPolicyOptions.MaxSurvivors.getValue();
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/UnalignedHeapChunk.java:
```diff
@@ -202,6 +202,18 @@ public static void dirtyCardForObjectOfUnalignedHeapChunk(Object obj, boolean ve
+     *
+     * This has to be fast, because it is used by the post-write barrier.
+     */
+    public static void dirtyCardForObjectOfUnalignedHeapChunk(Object obj) {
```
Here not calling `dirtyCardForObjectOfUnalignedHeapChunk(obj, false)` is because that method has the verify code: `AssertionNode.assertion(false, CardTable.isDirtyEntryAtIndexUnchecked...` , and `AssertionNode` will be called at compile time.
If we call this at GC will get exception here, could I just change this AssertionNode or implement a no verify one

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -74,6 +88,20 @@ public Log report(Log log, boolean traceHeapChunks) {
         log.string("[Young generation: ").indent(true);
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -140,4 +230,86 @@ boolean slowlyFindPointer(Pointer p) {
+        return toSpace.promoteAlignedObject(original);
+    }
+
+    private Object promoteUnAlignedObject(Object original, Space originalSpace) {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/heap/DiscoverableReference.java:
```diff
@@ -228,4 +228,8 @@ public static DiscoverableReference getNextDiscoverableReference(DiscoverableRef
         }
     }
+
+    public Pointer getNextRefPointer() {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -86,18 +114,80 @@ public final Space getSpace() {
+        Object result = null;
+        if (ObjectHeaderImpl.getObjectHeaderImpl().isAlignedObject(original)) {
+            trace.string("  aligned header: ").hex(ObjectHeaderImpl.readHeaderFromObject(original)).newline();
+            /* Promote by Object copying to the old generation. */
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapPolicyOptions.java:
```diff
@@ -74,4 +74,13 @@
+    public static final HostedOptionKey<Integer> MaxSurvivors = new HostedOptionKey<>(15);
+
+    @Option(help = "Number of survivor spaces") //
+    public static final RuntimeOptionKey<Integer> NumberOfSurvivors = new RuntimeOptionKey<>(4);
```
Yes, we make it a `RuntimeOptionKey` for not build the image again(which could be slow at sometime), the garbage collections can happen before this value gets parsed? We just follow the the keys like `MaximumYoungGenerationSizePercent`.
I'm not sure if this can create issues here. Is there a better way to change the value as the options of image runtime?
If this can create issues, I can change the related options as the hosted option

christianhaeubl replied:
With the code as it is now, `NumberOfSurvivorSpaces` will have to be a `HostedOptionKey`.

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/GCImpl.java:
```diff
@@ -131,6 +131,8 @@
     private boolean completeCollection;
     private UnsignedWord sizeBefore;
 
+    private final HeapVerifierImpl verifier;
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -140,4 +230,86 @@ boolean slowlyFindPointer(Pointer p) {
+
+    void prepareForPromotion() {
+        /* Prepare the Space walkers. */
+        for (int i = 0; i < HeapPolicyOptions.MaxSurvivors.getValue(); i++) {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/Space.java:
```diff
@@ -125,6 +134,17 @@ protected Space(String name, boolean isYoungSpace) {
+     * Creates the Survivor Space with age information during the compile-time.
+     */
+    @Platforms(Platform.HOSTED_ONLY.class)
+    protected Space(String name, boolean isFrom, int age) {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/Space.java:
```diff
@@ -136,6 +156,37 @@ final boolean isYoungSpace() {
+        return isSurvivorSpace;
+    }
+
+    Space asSurvivorSpace() {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapPolicyOptions.java:
```diff
@@ -74,4 +74,13 @@
     public static final RuntimeOptionKey<Long> UserRequestedGCThreshold = new RuntimeOptionKey<>(16L * 1024L * 1024L);
+
+    @Option(help = "Maximum value for survivor space") //
+    public static final HostedOptionKey<Integer> MaxSurvivors = new HostedOptionKey<>(15);
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/OldGeneration.java:
```diff
@@ -333,4 +353,8 @@ boolean walkHeapChunks(MemoryWalker.Visitor visitor) {
     }
+
+    boolean isOldSpace(Space thatSpace) {
+        return thatSpace == fromSpace || thatSpace == toSpace;
```
done, will determine the space eden/survivor/old  by `age`

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/AlignedHeapChunk.java:
```diff
@@ -263,6 +263,19 @@ public static void dirtyCardForObjectOfAlignedHeapChunk(Object object, boolean v
+     *
+     * This has to be fast, because it is used by the post-write barrier.
+     */
+    public static void dirtyCardForObjectOfAlignedHeapChunk(Object object) {
```
Yes, it seems duplicate code here. We want to keep one method too.
But if we call this method at runtime, will get exception of the `AssertionNode` at the image build time.

What's your suggestion about this. : )

below is the exception:
`
Fatal error: com.oracle.graal.pointsto.util.AnalysisError$ParsingError: Error encountered while parsing org.graalvm.compiler.replacements.nodes.AssertionNode.assertion(boolean, boolean, java.lang.String, java.lang.Object, java.lang.Object)
Parsing context:
	parsing org.graalvm.compiler.replacements.nodes.AssertionNode.assertion(AssertionNode.java:119)
	parsing com.oracle.svm.core.genscavenge.AlignedHeapChunk.dirtyCardForObjectOfAlignedHeapChunk(AlignedHeapChunk.java:261)
`

christianhaeubl replied:
You can avoid the error (and therefore the code duplication) if you change the call to `AssertionNode.assertion(...)` from
`AssertionNode.assertion(false, CardTable.isDirtyEntryAtIndexUnchecked(cardTableStart, index), "card must be dirty");`
to
`AssertionNode.assertion(false, CardTable.isDirtyEntryAtIndexUnchecked(cardTableStart, index), "card must be dirty", "", "");`

Roughly the same change can also be done for the unaligned chunks to avoid the code duplication there as well.

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/CollectionPolicy.java:
```diff
@@ -191,8 +191,9 @@ private static boolean collectCompletelyBasedOnTime(Log trace) {
+            final UnsignedWord youngInUse = getAccounting().getYoungChunkBytesAfter();
             final UnsignedWord oldInUse = getAccounting().getOldGenerationAfterChunkBytes();
-            final UnsignedWord withFullPromotion = youngSize.add(oldInUse).add(youngSize);
+            final UnsignedWord withFullPromotion = youngSize.add(oldInUse).add(youngSize).add(youngInUse);
```
Yes, you are right, the old code is sufficient to do the policy

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/GCImpl.java:
```diff
@@ -1375,17 +1439,20 @@ void afterCollectionCommon() {
             final UnsignedWord collectedChunkBytes = beforeChunkBytes.subtract(afterChunkBytes);
             collectedTotalChunkBytes = collectedTotalChunkBytes.add(collectedChunkBytes);
             if (HeapOptions.PrintGCSummary.getValue()) {
                 /* The young generation is empty after the collection. */
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/GCImpl.java:
```diff
@@ -1606,7 +1673,7 @@ private void printGCSummary() {
         JavaVMOperation.enqueueBlockingSafepoint("PrintGCSummaryShutdownHook", ThreadLocalAllocation::disableThreadLocalAllocation);
         final HeapImpl heap = HeapImpl.getHeapImpl();
-        final Space youngSpace = heap.getYoungGeneration().getSpace();
+        final Space youngSpace = heap.getYoungGeneration().getEden();
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapImpl.java:
```diff
@@ -286,7 +318,7 @@ boolean hasSurvivedThisCollection(Object obj) {
             final Space space = chunk.getSpace();
             final OldGeneration oldGen = getOldGeneration();
-            return space == oldGen.getToSpace();
+            return (space == oldGen.getToSpace() || (space.isSurvivorSpace() && !space.isFrom()));
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapImpl.java:
```diff
@@ -353,8 +389,8 @@ UnsignedWord getUsedChunkBytes() {
     UnsignedWord getYoungUsedChunkBytes() {
-        final Space.Accounting young = getYoungGeneration().getSpace().getAccounting();
-        return young.getAlignedChunkBytes().add(young.getUnalignedChunkBytes());
+        final Space.Accounting young = getYoungGeneration().getEden().getAccounting();
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapImpl.java:
```diff
@@ -377,7 +413,7 @@ UnsignedWord getOldUsedChunkBytes() {
     /** Return the size, in bytes, of the actual used memory, not the committed memory. */
     public UnsignedWord getUsedObjectBytes() {
-        final Space youngSpace = getYoungGeneration().getSpace();
+        final Space youngSpace = getYoungGeneration().getEden();
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/Space.java:
```diff
@@ -770,13 +807,13 @@ UnsignedWord getChunkBytes() {
 
     /** Aggregate the bytes in aligned chunks. */
-    private UnsignedWord getAlignedChunkBytes() {
+    UnsignedWord getAlignedChunkBytes() {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapImpl.java:
```diff
@@ -353,8 +389,8 @@ UnsignedWord getUsedChunkBytes() {
-        final Space.Accounting young = getYoungGeneration().getSpace().getAccounting();
-        return young.getAlignedChunkBytes().add(young.getUnalignedChunkBytes());
+        final Space.Accounting young = getYoungGeneration().getEden().getAccounting();
+        return young.getAlignedChunkBytes().add(young.getUnalignedChunkBytes()).add(getYoungGeneration().getChunkUsedBytes());
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapPolicy.java:
```diff
@@ -455,8 +465,9 @@ public static Sometimes factory() {
+                final UnsignedWord youngInUse = heap.getYoungUsedChunkBytes();
                 /* Has there been enough allocation to provoke a collection? */
-                if (bytesAllocatedSinceLastCollection.get().aboveOrEqual(getMaximumYoungGenerationSize())) {
+                if (bytesAllocatedSinceLastCollection.get().aboveOrEqual(getMaximumYoungGenerationSize().subtract(youngInUse))) {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/OldGeneration.java:
```diff
@@ -118,7 +122,7 @@ private boolean shouldPromoteFrom(Space originalSpace) {
         final HeapImpl heap = HeapImpl.getHeapImpl();
         if (heap.isYoungGeneration(originalSpace)) {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/HeapPolicyOptions.java:
```diff
@@ -74,4 +74,11 @@
+    public static final HostedOptionKey<Integer> MaxSurvivorSpaces = new HostedOptionKey<>(15);
+
+    @Option(help = "Number of survivor spaces") //
+    public static final RuntimeOptionKey<Integer> NumberOfSurvivorSpaces = new RuntimeOptionKey<>(4);
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -112,7 +197,7 @@ protected boolean verify(final HeapVerifierImpl.Occasion occasion) {
         final HeapImpl heap = HeapImpl.getHeapImpl();
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -32,48 +33,72 @@
 
     /** Return all allocated virtual memory chunks to HeapChunkProvider. */
     @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
     public final void tearDown() {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -32,48 +33,72 @@
     }
 
     @Override
     public boolean walkObjects(ObjectVisitor visitor) {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -130,14 +215,129 @@ protected boolean verify(final HeapVerifierImpl.Occasion occasion) {
 
     boolean slowlyFindPointer(Pointer p) {
-        if (HeapVerifierImpl.slowlyFindPointerInSpace(getSpace(), p, HeapVerifierImpl.ChunkLimit.top)) {
+        if (HeapVerifierImpl.slowlyFindPointerInSpace(getEden(), p, HeapVerifierImpl.ChunkLimit.top)) {
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -130,14 +215,129 @@ protected boolean verify(final HeapVerifierImpl.Occasion occasion) {
         return false;
     }
 
     boolean walkHeapChunks(MemoryWalker.Visitor visitor) {
```
done

christianhaeubl replied:
At the moment, you forcibly walk all spaces, even if the visitor wants to stop visiting (i.e., if `walkHeapChunks(...)` returns `false`).

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -130,14 +215,129 @@ protected boolean verify(final HeapVerifierImpl.Occasion occasion) {
+    }
+
+    private static boolean shouldPromoteFrom(Space originalSpace) {
+        return !originalSpace.isSurvivorSpace() || originalSpace.isFrom();
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -130,14 +215,129 @@ protected boolean verify(final HeapVerifierImpl.Occasion occasion) {
+        if (!shouldPromoteFrom(originalSpace)) {
+            return original;
+        }
+        int age = originalSpace.isEdenSpace() ? 1 : originalSpace.getNextAgeForPromotion();
```
done

tbzhang commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/YoungGeneration.java:
```diff
@@ -130,14 +215,129 @@ protected boolean verify(final HeapVerifierImpl.Occasion occasion) {
+        if (shouldPromoteFrom(originalSpace)) {
+            trace.string("  promoting");
+            final UnalignedHeapChunk.UnalignedHeader uChunk = UnalignedHeapChunk.getEnclosingUnalignedHeapChunk(original);
+            int age = originalSpace.isEdenSpace() ? 1 : originalSpace.getNextAgeForPromotion();
```
done

tbzhang commented:
@christianhaeubl Thank you for your review comments and I look forward to your new feedback. Meanwhile, I wonder if I need to solve the "travis-ci” problem to make it passed

christianhaeubl commented:
Thanks for doing the changes. I added one more minor comment. Besides that, the changes look good now. I will run our internal checks and let you know if any further changes are needed.

Regarding "travis-ci": you should fix the code formatting issue that was found by one of the failed Travis jobs.

Regarding the performance results that you mentioned in your very first comment: did the results change now that the size of the young generation is limited correctly?

tbzhang commented:
> Thanks for doing the changes. I added one more minor comment. Besides that, the changes look good now. I will run our internal checks and let you know if any further changes are needed.
> 
> Regarding "travis-ci": you should fix the code formatting issue that was found by one of the failed Travis jobs.
> 
> Regarding the performance results that you mentioned in your very first comment: did the results change now that the size of the young generation is limited correctly?

Thanks for your reply, I fixed that comment you mentioned.

Also, I fix that "travis-ci" code formatting issue, I will follow the Travis jobs

I tested the build at each update of this PR, and didn't see the significant regression of the frequency of FGCs. the previous size of young generation didn't calculate the survivor used space into the young generation size. After that fix, it will increase the frequency of incremental gc, But in our test, the usage space of survivor space didn't take too much of young size: ~5%. So this fix didn't change the performance result a lot

christianhaeubl commented:
Thanks. I started the process of integrating your PR into master. Our internal checks revealed a couple more issues but I decided to directly fix those in separate commits on top of your PR. If everything goes as expected, your PR will get merged into master in the next couple of days.

adinn commented:
> Thanks. I started the process of integrating your PR into master

Hurrah!

tbzhang commented:
> Thanks. I started the process of integrating your PR into master. Our internal checks revealed a couple more issues but I decided to directly fix those in separate commits on top of your PR. If everything goes as expected, your PR will get merged into master in the next couple of days.

That's great. Thank you!

fniephaus commented:
Fantastic, thanks for the contribution! 🚀

christianhaeubl commented:
This PR caused quite a significant slow-down in many of our internal benchmarks as it does not work particularly well with the current default heap policy that tries to balance the time spent in young vs. old GCs: survivor spaces will increase the time spent in young GCs, which then also increases the number of full GCs.

So, as part of a change that improves overall GC performance significantly, I also changed the default from `-H:MaxSurvivorSpaces=4` to `-H:MaxSurvivorSpaces=0`. This effectively disables survivor spaces by default until we have a better heap policy.
